File: admin-ui/src/api/tenants.ts
import client from './client';

export interface Tenant {
  id: string;
  name: string;
  plan: string;
  is_active: boolean;
  created_at: string;
}

export interface TenantQuota {
  max_documents: number;
  max_storage_bytes: number;
  max_queries_per_day: number;
  max_users: number;
  max_workspaces: number;
}

export interface TenantUsage {
  current_documents: number;
  current_storage_bytes: number;
  queries_today: number;
  current_users: number;
  current_workspaces: number;
}

interface ListResponse<T> {
  data: T[];
  total: number;
}

export async function listTenants(): Promise<Tenant[]> {
  const res = await client.get<ListResponse<Tenant>>('/api/km/tenants');
  return res.data.data;
}

export async function createTenant(name: string, plan: string): Promise<Tenant> {
  const res = await client.post<Tenant>('/api/km/tenants', { name, plan });
  return res.data;
}

export async function getTenant(id: string): Promise<Tenant> {
  const res = await client.get<Tenant>(`/api/km/tenants/${id}`);
  return res.data;
}

export async function updateTenant(id: string, name: string, plan: string): Promise<Tenant> {
  const res = await client.put<Tenant>(`/api/km/tenants/${id}`, { name, plan });
  return res.data;
}

export async function deleteTenant(id: string): Promise<void> {
  await client.delete(`/api/km/tenants/${id}`);
}

export async function getTenantQuota(id: string): Promise<TenantQuota> {
  const res = await client.get<TenantQuota>(`/api/km/tenants/${id}/quota`);
  return res.data;
}

export async function setTenantQuota(id: string, quota: TenantQuota): Promise<void> {
  await client.put(`/api/km/tenants/${id}/quota`, quota);
}

export async function getTenantUsage(id: string): Promise<TenantUsage> {
  const res = await client.get<TenantUsage>(`/api/km/tenants/${id}/usage`);
  return res.data;
}

export async function assignOrg(tenantId: string, orgId: string): Promise<void> {
  await client.post(`/api/km/tenants/${tenantId}/assign-org`, { org_id: orgId });
}

File: crates/thairag-api/src/routes/tenants.rs
use axum::Json;
use axum::extract::{Path, State};
use axum::http::StatusCode;
use serde::{Deserialize, Serialize};

use crate::app_state::AppState;
use crate::error::{ApiError, AppJson};
use crate::store::{Tenant, TenantQuota, TenantUsage};
use thairag_core::types::OrgId;

// ── DTOs ────────────────────────────────────────────────────────────

#[derive(Deserialize)]
pub struct CreateTenantRequest {
    pub name: String,
    pub plan: String,
}

#[derive(Deserialize)]
pub struct UpdateTenantRequest {
    pub name: String,
    pub plan: String,
}

#[derive(Deserialize)]
pub struct AssignOrgRequest {
    pub org_id: uuid::Uuid,
}

#[derive(Serialize)]
pub struct ListResponse<T: Serialize> {
    pub data: Vec<T>,
    pub total: usize,
}

// ── Handlers ─────────────────────────────────────────────────────────

pub async fn create_tenant(
    State(state): State<AppState>,
    AppJson(body): AppJson<CreateTenantRequest>,
) -> Result<(StatusCode, Json<Tenant>), ApiError> {
    let tenant = state.km_store.insert_tenant(body.name, body.plan)?;
    Ok((StatusCode::CREATED, Json(tenant)))
}

pub async fn list_tenants(State(state): State<AppState>) -> Json<ListResponse<Tenant>> {
    let data = state.km_store.list_tenants();
    let total = data.len();
    Json(ListResponse { data, total })
}

pub async fn get_tenant(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Result<Json<Tenant>, ApiError> {
    let tenant = state.km_store.get_tenant(&id)?;
    Ok(Json(tenant))
}

pub async fn update_tenant(
    State(state): State<AppState>,
    Path(id): Path<String>,
    AppJson(body): AppJson<UpdateTenantRequest>,
) -> Result<Json<Tenant>, ApiError> {
    let tenant = state.km_store.update_tenant(&id, body.name, body.plan)?;
    Ok(Json(tenant))
}

pub async fn delete_tenant(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Result<StatusCode, ApiError> {
    state.km_store.delete_tenant(&id)?;
    Ok(StatusCode::NO_CONTENT)
}

pub async fn get_quota(State(state): State<AppState>, Path(id): Path<String>) -> Json<TenantQuota> {
    let quota = state.km_store.get_tenant_quota(&id);
    Json(quota)
}

pub async fn set_quota(
    State(state): State<AppState>,
    Path(id): Path<String>,
    AppJson(quota): AppJson<TenantQuota>,
) -> Result<Json<TenantQuota>, ApiError> {
    state.km_store.set_tenant_quota(&id, &quota)?;
    Ok(Json(quota))
}

pub async fn get_usage(State(state): State<AppState>, Path(id): Path<String>) -> Json<TenantUsage> {
    let usage = state.km_store.get_tenant_usage(&id);
    Json(usage)
}

pub async fn assign_org(
    State(state): State<AppState>,
    Path(id): Path<String>,
    AppJson(body): AppJson<AssignOrgRequest>,
) -> Result<StatusCode, ApiError> {
    state
        .km_store
        .assign_org_to_tenant(OrgId(body.org_id), &id)?;
    Ok(StatusCode::NO_CONTENT)
}
